onnxruntime/core/framework/execution_provider.cc

```diff
@@ -52,103 +52,64 @@
 #endif
 }
 
-// Returns true if an allocator was found and replaced
-static bool FindAndReplaceAllocator(const OrtMemoryInfo& mem_info,
-                                    const MemoryInfoSet& mem_info_set,
-                                    AllocatorMap& allocators,
-                                    AllocatorPtr replacing_allocator) {
-  auto ite = mem_info_set.find(mem_info);
-
-  if (ite != mem_info_set.end()) {
-    const int key = MakeKey(mem_info.id, mem_info.mem_type);
-    allocators[key] = replacing_allocator;
-    return true;
-  }
-
-  return false;
-}
-
 // Update allocator in the provider if already present; ignore if not.
+// We match using the device id, OrtMemType and OrtDevice info.
+// We ignore the allocator name, and OrtAllocatorType (whether internally an arena is used or not).
+// TODO: We should remove OrtAllocatorType from OrtMemoryInfo as it's an implmentation detail of the allocator.
 void IExecutionProvider::ReplaceAllocator(AllocatorPtr allocator) {
```

Comment:
@pranavsharma let me know if you think there are any issues with this more OrtDevice centric approach.